src/write_resume_data.cpp

```diff
@@ -154,9 +109,161 @@ namespace libtorrent {
 							ret_mask.push_back(bit ? '1' : '0');
 					}
 				}
+
+				if (!(flags & write_resume_only) && atp.ti)
+				{
+					file_storage const& fs = atp.ti->files();
+					if (!fs.pad_file_at(f) && fs.file_size(f) > fs.piece_length())
+					{
+						aux::merkle_tree t(fs.file_num_blocks(f)
+							, fs.piece_length() / default_block_size, fs.root_ptr(f));
+
+						if (f < atp.merkle_tree_mask.end_index() && !atp.merkle_tree_mask[f].empty())
+						{
+							t.load_sparse_tree(tree, atp.merkle_tree_mask[f]);
+						}
+						else
+						{
+							t.load_tree(tree);
+						}
+
+						auto const piece_layer = t.get_piece_layer();
+						auto& layer = piece_layers[t.root().to_string()].string();
+
+						for (auto const& h : piece_layer)
+						{
+							layer += h.to_string();
+						}
+					}
+				}
+			}
+
+			if (!piece_layers.empty())
+			{
+				ret["piece layers"] = piece_layers;
 			}
 		}
 
+		// save web seeds
+		if (!atp.url_seeds.empty())
+		{
+			entry::list_type& url_list = ret["url-list"].list();
+			std::copy(atp.url_seeds.begin(), atp.url_seeds.end(), std::back_inserter(url_list));
+		}
+
+		if (!atp.http_seeds.empty())
+		{
+			entry::list_type& httpseeds_list = ret["httpseeds"].list();
+			std::copy(atp.http_seeds.begin(), atp.http_seeds.end(), std::back_inserter(httpseeds_list));
+		}
+
+		if (!atp.name.empty()) ret["name"] = atp.name;
+
+		// save trackers
+		entry announce_list;
+		if (!atp.trackers.empty())
+		{
+			if (atp.trackers.size() == 1 && !(flags & write_resume_only))
+			{
+				ret["announce"] = atp.trackers.front();
+			}
+
+			if (!(flags & write_torrent_only))
+			{
+				entry::list_type& tr_list = announce_list.list();
+				tr_list.emplace_back(entry::list_type());
+				std::size_t tier = 0;
+				auto tier_it = atp.tracker_tiers.begin();
+				for (std::string const& tr : atp.trackers)
+				{
+					if (tier_it != atp.tracker_tiers.end())
+						tier = aux::clamp(std::size_t(*tier_it++), std::size_t{0}, std::size_t{1024});
+
+					if (tr_list.size() <= tier)
+						tr_list.resize(tier + 1);
+
+					tr_list[tier].list().emplace_back(tr);
+				}
+			}
+
+			if (!(flags & write_resume_only) && atp.trackers.size() > 1)
+				ret["announce-list"] = announce_list;
+		}
+
+		// .torrent file fields above
+
+		if (flags & write_torrent_only) return ret;
```

Comment:
Nitpick
I would implement it in a reversed way:
1. Function that (really) produces torrent data
2. Function that (really) produces resume data
3. Function that produces merged (or one of, depending of parameter) data above.